code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,461 @@
+			icon_slots[3] = rand(1, 2)
+		generate_icon(icon, "IL[icon_slots[3]]")
+		generate_icon(icon, "ILL[icon_slots[3]]", material)
+
+	if(prob(50)  || icon_slots[4])//Right
+		if(!(icon_slots[4]))
+			icon_slots[4] = rand(1, 2)
+		generate_icon(icon, "IR[icon_slots[4]]")
+		generate_icon(icon, "IRL[icon_slots[4]]", material)
+
+	if(prob(30) || icon_slots[2])//Bottom
+		if(!(icon_slots[2]))
+			icon_slots[2] = rand(1, 2)
+		generate_icon(icon, "IBTM[icon_slots[2]]")
+		generate_icon(icon, "IBTML[icon_slots[2]]", material)
+
+/obj/item/xenoartifact/CanAllowThrough(atom/movable/mover, turf/target) //tweedle dee, density feature
+	. = ..()
+	if(get_trait(/datum/xenoartifact_trait/minor/dense))
+		return FALSE
+
+/obj/item/xenoartifact/attack_hand(mob/user) //tweedle dum, density feature
+	if(get_trait(/datum/xenoartifact_trait/minor/dense))
+		if(user.a_intent != INTENT_GRAB)
+			SEND_SIGNAL(src, XENOA_INTERACT, null, user, user) //Calling the regular attack_hand signal causes feature issues, like picking up the artifact.
+		else
+			touch_desc?.on_touch(src, user)
+		return FALSE
+	..()
+
+/obj/item/xenoartifact/examine(mob/living/carbon/user)
+	. = ..()
+	if(isobserver(user))
+		to_chat(user, "<span class='notice'>[special_desc]</span>")
+		for(var/datum/xenoartifact_trait/t as() in traits)
+			to_chat(user, "<span class='notice'>[t.desc ? t.desc : t.label_name]\n</span>")
+	else if(iscarbon(user) && istype(user?.glasses, /obj/item/clothing/glasses/science)) //Not checking carbon throws a runtime concerning observers
+		to_chat(user, "<span class='notice'>[special_desc]</span>")
+
+/obj/item/xenoartifact/interact(mob/user)
+	. = ..()
+	if(process_type == IS_LIT) //Snuff out candle
+		to_chat(user, "<span class='notice'>You snuff out [name]</span>")
+		process_type = null
+		return
+	if(user.a_intent == INTENT_GRAB)
+		touch_desc?.on_touch(src, user)
+		return
+	SEND_SIGNAL(src, XENOA_INTERACT, null, user, user)
+
+/obj/item/xenoartifact/attackby(obj/item/I, mob/living/user, params)
+	for(var/datum/xenoartifact_trait/t as() in traits)
+		t.on_item(src, user, I)
+	if(!(COOLDOWN_FINISHED(src, xenoa_cooldown))||user?.a_intent == INTENT_GRAB||istype(I, /obj/item/xenoartifact_label)||istype(I, /obj/item/xenoartifact_labeler))
+		return
+	..()
+
+/obj/item/xenoartifact/proc/check_charge(mob/user, charge_mod) //Run traits
+	if(logging)
+		log_game("[user] attempted to activate [src] at [world.time]. Located at [x] [y] [z].")
+
+	if(COOLDOWN_FINISHED(src, xenoa_cooldown))
+		if(prob(malfunction_chance)) //See if we pick up an malfunction
+			var/datum/xenoartifact_trait/t = pick(subtypesof(/datum/xenoartifact_trait/malfunction))
+			traits+=new t
+			malfunction_chance=malfunction_chance*0.2 //Lower chance after contracting 
+		else    
+			malfunction_chance+=malfunction_mod //otherwise increase chance
+
+		for(var/atom/M in true_target) //Cull bad targets
+			if(get_dist(get_turf(src), get_turf(M)) > max_range)   
+				true_target -= M
```

Comment:
this loop isn't even needed. If the target is out of range you can remove it in the second loop and continue